### ISSUE TITLE
binary crossentropy loss function not supported?

### ISSUE THREAD

**Issue by ChristopherPS96** [SciSharp/TensorFlow.NET#832]
Hello,
just three short question:
1. Is binary_crossentropy not supported yet? Can't find it under keras.losses. and could really need it for my multilabel problem
2. Can I implement it myself? Through extending from ILossFunc with my own custom class?
3. Does Keras.NET has this loss function? But after reading through the github page, I dont think this and Keras.NET doesnt work together.

Thanks for the help 👍

**Comment by Oceania2018** [comment]
https://github.com/SciSharp/SciSharp-Stack-Examples/blob/3f1713cdb4ae033878978224c3921adfa680c746/src/TensorFlowNET.Examples/GAN/MnistGAN.cs#L182

**Comment by ismasanchez** [comment]
I could resolve this issue in two ways:

Option 1: Using CategoricalCrossentropy Loss function instead.

In this option, the target that is a boolean 0 or 1, is separated into an array like this: 0 => {0,1} and 1 => {1,0}

The last layers should then be like this: keras.layers.Dense(2, activation: keras.activations.Softmax).Apply(outputs)

model.compile(loss: keras.losses.CategoricalCrossentropy(), optimizer: keras.optimizers.Adam(), metrics: new[] { "acc" });

Option 2: Applying binary crossentropy by creating your own class.

    public class BinaryCrossentropy : LossFunctionWrapper, ILossFunc
    {
        private float label_smoothing;

        public BinaryCrossentropy(bool from_logits = false, float label_smoothing = 0f, string reduction = null, string name = null)
            : base(reduction, (name == null) ? "binary_crossentropy" : name, from_logits)
        {
            this.label_smoothing = label_smoothing;
        }

        public override Tensor Apply(Tensor y_true, Tensor y_pred, bool from_logits = false, int axis = -1)
        {
            return BinaryCrossentropyLoss(y_true, y_pred);
        }
 
        private Tensor BinaryCrossentropyLoss(Tensor x, Tensor y)
        {
            var shape = tf.reduce_prod(tf.shape(x));
            var count = tf.cast(shape, TF_DataType.TF_FLOAT);
            x = tf.clip_by_value(x, 1e-6f, 1.0f - 1e-6f);
            var z = y * tf.log(x) + (1 - y) * tf.log(1 - x);
            var result = -1.0f / count * tf.reduce_sum(z);
            return result;
        }
    }

And then you use it like this:
model.compile(loss: new BinaryCrossentropy(), optimizer: keras.optimizers.Adam(), metrics: new[] { "acc" });

Please note with this implementation I'm not 100% confident that is absolutely correct, it would be nice if someone else can confirm.
